### PR TITLE
Openst setup

### LINKED ISSUE #32: Setup of OpenST
- Setup of master copies
 	- Deployment of CreateAndAddModules contract.
 	- Setup of factory contracts.
 		UserWalletFactory contract
 		ProxyFactory contract
       - Deployment of TokenRules.
       - Deployment of PricerRule.
       - Store the addresses in config.json file.
       - Above will be stitched in openst.js and agent should call the setup.	

- Command:
 	setupOpenST config.json

### LINKED ISSUE #42: TokenRules deployment should also be part of OpenST Setup JLP
Currently TokenRules is a separate JLP command.

We should deploy TokenRules as part of openst setup. This way with a single JLP command economy setup can be done.

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers setup of openst using a single JLP command:

**Command:**
```
node ./src/bin/openst.js config.json organization eip20Token
```

- The command performs below tasks:
    Deployment of TH, GnosisSafe, Recovery master copies
    Deployment of CreateAndAddModules contract.
    Deployment of UserWalletFactory, ProxyFactory contracts.
    Deployment of TokenRules.

- Addresses are stored config.json file. 
- Readme has been updated with setup command.
- Currently openst.js npm points to https://github.com/abhayks1/openst.js.git#openst_setup since it's in PR. Once the openst.js PR is merged, It will be switched back to latest openst.js npm.
- For deployment of openst contracts openst_deployer.js has been added.

Fixes #32 
Fixes #42

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -209,26 +209,31 @@ node src/bin/facilitator.js progressRedeem path_to_config.json messageHash
...
+ * Recovery master copy address.
+ * UserWalletFactory contract address.
+ * ProxyFactory contract address.
+ * CreateAndAddModule helper contract address.
+ * TokenRules contract address.
+    
+ * Replace `config.json` with the path to the configuration file.
+ * Replace `organization` with an organization contract address. 
+ * Replace `eip20Token` with eip20Token address.
+
```
Comment (schemar) [inline README.md]: Why did you add a leading space to every line?

File: README.md [hunk 2]
```diff
@@ -252,5 +257,4 @@ Get the base token balance of a local account.
 node ./src/bin/balance.js --help
 
 # Deploy TokenRules token:
-node ./src/bin/balance.js config.json auxiliary
-```
+node ./src/bin/balance.js config.json auxiliary
```
Comment (schemar) [inline README.md]: Why did you remove the last line with the ending ` ``` `?

File: README.md [hunk 3]
```diff
@@ -209,26 +209,31 @@ node src/bin/facilitator.js progressRedeem path_to_config.json messageHash
...
-It will write `tokenRulesAddress` to your config file.
-
-* Replace `config.json` with the path to the configuration file.
-* Replace `eip20Token` with eip20Token address.
-* Replace `organization` with an organization contract address. 
-
+  * `messageHash` generated in redeem facilitator step.
+  
+ ## Setup OpenST
+
```
Comment (schemar) [inline README.md]: Are there no prerequisites for setting up OpenST?

File: src/bin/openst.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  .version(version)
+  .name('bt')
+  .arguments('openst <config> <organization> <eip20Token>')
+  .description('An executable to setup OpenST.')
+  .action(
+    async (config, organization, eip20Token) => {
+      const chainConfig = new ChainConfig(config);
+      const openstDeployer = new OpenSTDeployer(chainConfig);
+      await openstDeployer.setupOpenst(organization, eip20Token);
+      chainConfig.write(config);
```
Comment (schemar) [inline src/bin/openst.js]: Please use the `connected.run()` function to run executables that send transactions. See other executables as an example. Maybe you need to update your branch from `master`.

`connected.run()` assures you have the correct Web3 providers set up and the connection is closed afterwards.

File: src/config/chain_config.js [hunk 5]
```diff
@@ -7,10 +7,10 @@ class ChainConfig {
    * @param {string} [filePath] An optional config file to overwrite any defaults or env variables.
    */
   constructor(filePath) {
-    this.originWeb3Provider = process.env.ORIGIN_WEB3_PROVIDER || 'http://localhost:8545';
-    this.auxiliaryWeb3Provider = process.env.AUXILIARY_WEB3_PROVIDER || 'http://localhost:8547';
+    this.originWeb3Provider = process.env.ORIGIN_WEB3_PROVIDER || 'http://localhost:8546';
+    this.auxiliaryWeb3Provider = process.env.AUXILIARY_WEB3_PROVIDER || 'http://localhost:8545';
```
Comment (schemar) [inline src/config/chain_config.js]: Why?

File: src/openst_deployer.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
+const Web3 = require('web3');
+const Package = require('@openstfoundation/openst.js');
+
+const logger = require('./logger');
+
+class OpenSTDeployer {
```
Comment (schemar) [inline src/openst_deployer.js]: Why not just `OpenST` (and then `OpenST::deployXxx()`)

File: src/openst_deployer.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  constructor(chainConfig) {
+    this.chainConfig = chainConfig;
+    this.auxiliary = {
+      web3: new Web3(chainConfig.auxiliaryWeb3Provider),
```
Comment (schemar) [inline src/openst_deployer.js]: Web3 should be injected with the proper provider. It is probably beneficial to inject the `Connection` (See `connected.run()`). See the latest code of JLP for that.

File: README.md (hunk 1 above)
Comment (abhayks1) [reply README.md]: It's not needed. Removed it 👍

File: src/config/chain_config.js (hunk 5 above)
Comment (abhayks1) [reply src/config/chain_config.js]: 😞

File: src/openst_deployer.js (hunk 6 above)
Comment (abhayks1) [reply src/openst_deployer.js]: Good point 👍 Renaming done!

File: README.md (hunk 3 above)
Comment (abhayks1) [reply README.md]: Updated!

File: src/bin/openst.js (hunk 4 above)
Comment (abhayks1) [reply src/bin/openst.js]: Done!

File: README.md [hunk 8]
```diff
@@ -215,36 +215,42 @@ node src/bin/facilitator.js progressRedeem path_to_config.json messageHash
...
-  node ./src/bin/registerRule.js config.json ruleName ruleAddress ruleAbi
-  ```
+Prerequisite: `tokenRules` and `worker` in your config file.
+
+```bash
+# Help:
+node ./src/bin/registerRule.js --help
+
+# Register rule to TokenRules:
+node ./src/bin/registerRule.js config.json ruleName ruleAddress ruleAbi
```
Comment (gulshanvasnani) [inline README.md]: Try registering an rule with an ruleAbi.
